Load runtime version on runtime build

David-Petrov opened the pull request:
- add functionality to first try to parse the runtime version directly from the wasm blob's custom sections and if this fails (sections not present), fall back to calling `Core_version` on the built runtime. In this, fix the previously existing parsing from custom sections;

- move the essence of shared memory reading into the `Runtime` class (i.e. the method `getDataFromMemory`), was previously in `HostApi`

- use the state version field of the runtime version for the genesis trie

- add tests (actually, only a single working one, since the runtime builder tests turned out more complicated than expected)

Closes #290

Commit:
fix!: rename two imports

runtime is expecting them with those names, so it wasn't building

Commit:
feat!: fetch runtime version on runtime build

- add functionality to first try to parse the runtime version directly from the wasm blob and if this fails, fall back to calling `Core_version` on the built runtime (in this, fix the previously existing parsing from custom sections)

- move the essence of memory reading into the `Runtime` class (i.e. the method `getDataFromMemory`)

- use the state version field of the runtime version for the genesis trie

- add tests (actually, only a single working one, since the runtime builder tests turned out more complicated than expected)

David-Petrov commented on src/test/java/com/limechain/runtime/RuntimeBuilderTest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Test
+    @Disabled("Very lengthy, needs either refactoring in the code being tested or a lot of effort into mocking")
+    void buildsRuntimeWithVersionInCustomSection() throws IOException {
```
We might as well delete this test, since it'll be a cumbersome task to write properly.
Main issue is the dependency on the spring context which we have in the very constructor of a `Runtime` instance.
A few approach options are possible:
Ideally, we'd want an end-to-end test with the real blob, so we could:
1) try to mock out the entire `HostApi` implementations;
2) try to stub with an empty implementation of all imports;
3) give up on the end to end / integration test and simply test what's relevant to the `RuntimeBuilder` (which is in sync with unit testing, but we lose the actual check that our logic works with a real blob)

The frustrating part is that the currently written logic actually works, but for now only inspected whilst debugging in production... some kind of automation and a regression test that we won't break it later would be very fit at this point.

georg-getz replied:
Have you tried using the runtime that's in the examples folder of wasmer-java? It is very lightweight and only requires 2 function imports and a memory import.

David-Petrov replied:
In a side discussion, we decided to scrape the test for now as out of scope. Added to `TODO.md` instead.

georg-getz commented on src/main/java/com/limechain/runtime/WasmSectionUtils.java:
```diff
@@ -0,0 +1,124 @@
+import java.util.Arrays;
+import java.util.Objects;
+
+// TODO: [Testability concern] Maybe make this class non-static?
```
But it already is non static, do you mean removing the `@UtilityClass` annotation?

David-Petrov replied:
Yes, exactly. The two opposing points are:
1) pro: static methods are more cumbersome to mock; mocking an instance is easier (in tests, which don't exist yet)
2) con: logically, it's a static utility class (by its purpose), so instantiating it everywhere might be considered "annoying boilerplate" or "unnecessary" since it doesn't hold any significant state.

georg-getz replied:
I prefer leaving it in as changing code for ease of testing isn't always a good practice.

georg-getz commented on src/main/java/com/limechain/runtime/WasmSectionUtils.java:
```diff
@@ -0,0 +1,124 @@
+     */
+    @Nullable
+    public RuntimeVersion parseRuntimeVersionFromCustomSections(byte[] wasmBinary) {
+        // byte value of \0asm concatenated with 0x1, 0x0, 0x0, 0x0 from smoldot runtime_version.rs#97
```
Not a good practice to point to a concrete line in another repo since it can easily be changed making this comment false.

David-Petrov replied:
I agree, but I left this comment as it was since I was unable to find anything regarding that prefix in the spec / anywhere in an official source... I'm most probably missing something, but alternatively, we could link to a certain commit hash in the smoldot repo in order to not lose the motivation of this "magic prefix" 😄

georg-getz replied:
commit hash is a bit better for sure

georg-getz commented on src/main/java/com/limechain/runtime/WasmSectionUtils.java:
```diff
@@ -0,0 +1,124 @@
+            int sectionSize = 0;
+            int shift = 0;
+            byte byteRead;
+            do {
+                byteRead = wasmBytes[offset++];
+                sectionSize |= (byteRead & 0x7f) << shift;
+                shift += 7;
+            } while (byteRead < 0);
```
Can you add a small comment explaining this do while? For example where is the 7 coming from?

David-Petrov replied:
Again, that was magic to myself, too... I'm not sure where this comes from as I've not investigated it in depth, probably the reason I was hesitant to change it in the slightest 😄

georg-getz commented on src/main/java/com/limechain/runtime/version/ApiVersions.java:
```diff
@@ -0,0 +1,66 @@
+
+        //TODO: Think about checking if the length is divisible by the fixed byte size of an encoded api version
+        while (reader.hasNext()) {
+            try {
+                apiVersionsRaw.add(reader.read(apiVersionReader));
+            } catch (RuntimeException e) {
+                break; // Presume end of iteration (all is ok), although other exceptions are possible
+            }
```
It's probably good to have a log here, if the exception is expected the log could be set to a lower logging level, but it's good to have some info about the exception.

David-Petrov replied:
Actually, an exception was not expected per se, it's rather really an exceptional scenario, it was a miss on my part.
Take a look again if we're satisfied with this handling (javadoc improved as well): a037d61

Commit:
chore: add `toString()` in `ApiVersion`

Commit:
chore: remove dysfunctional test and introduce a TODO.md instead

- `TODO.md` is planned to grow into a backlog for future out-of-scope necessities stumbled upon during development

Commit:
chore: add past concern to `TODO.md`

Commit:
chore: remove testability concern TODO

Commit:
chore: improving comment pointing to Smoldot

Commit:
chore: add javadoc and improve exception handling in `ApiVersions.decodeNoLength`

- to better handle and emphasize the peculiarity of the missing list length in the scale encoded input

Commit:
chore(sonarlint): add a `toString()` to `WasmCustomSection` record

simply for reliability and consistency purposes with Java's requirements, not that anyone would have significant use of its presence (perhaps in rare debugging cases, printing opaque byte sections from the wasm blob)

Commit:
refactor: change method name

currently unused (was used), but makes sense to exist nonetheless, so no breaking changes introduced
